### PR TITLE
Read file length for symlinks,

### PR DESCRIPTION
**Pull request by cvium**
**Changes**
Created a subclass of `PhysicalFileResultExecutor` and replaced it in the DI container. It unfortunately reads the file attributes 1-2 times more than its superclass, but `FileInfo` is called multiple times even outside `PhysicalFileResultExecutor`. Therefore I have copied parts of static extensions methods eg. https://github.com/dotnet/aspnetcore/blob/main/src/Http/Http/src/SendFileFallback.cs to get the desired result.

The new executor is only used for symlinks.

Supersedes #5775 and #5824. And a huge thanks for the research done by both PR authors.

**Issues**
Fixes https://github.com/jellyfin/jellyfin/issues/5521

### CODE REVIEW

File: Jellyfin.Server/Infrastructure/SymlinkFollowingPhysicalFileResultExecutor.cs [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            }
+
+            if (count.HasValue
+                && (count.Value < 0 || count.Value > fileInfo.Length - offset))
+            {
+                throw new ArgumentOutOfRangeException(nameof(count), count, string.Empty);
+            }
+
+            // Copied from SendFileFallback.SendFileAsync
+            const int bufferSize = 1024 * 16;
```
Comment (Bond-009) [inline Jellyfin.Server/Infrastructure/SymlinkFollowingPhysicalFileResultExecutor.cs:127]:
```suggestion
            const int BufferSize = 1024 * 16;
```

File: Jellyfin.Server/Infrastructure/SymlinkFollowingPhysicalFileResultExecutor.cs [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+            var fileInfo = GetFileInfo(filePath);
+            if (offset < 0 || offset > fileInfo.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset), offset, string.Empty);
+            }
+
+            if (count.HasValue
+                && (count.Value < 0 || count.Value > fileInfo.Length - offset))
+            {
+                throw new ArgumentOutOfRangeException(nameof(count), count, string.Empty);
```
Comment (Bond-009) [inline Jellyfin.Server/Infrastructure/SymlinkFollowingPhysicalFileResultExecutor.cs:123]: Pass a message?

**Review by Bond-009 — APPROVED** [review]

File: Jellyfin.Server/Infrastructure/SymlinkFollowingPhysicalFileResultExecutor.cs (hunk 2 above)
Comment (cvium) [reply Jellyfin.Server/Infrastructure/SymlinkFollowingPhysicalFileResultExecutor.cs:123]: I suppose. This is just copied from aspnetcore

File: Jellyfin.Server/Infrastructure/SymlinkFollowingPhysicalFileResultExecutor.cs [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+            // Copied from SendFileFallback.SendFileAsync
+            const int BufferSize = 1024 * 16;
+
+            await using var fileStream = new FileStream(
+                filePath,
+                FileMode.Open,
+                FileAccess.Read,
+                FileShare.ReadWrite,
+                bufferSize: bufferSize,
```
Comment (cvium) [inline Jellyfin.Server/Infrastructure/SymlinkFollowingPhysicalFileResultExecutor.cs:134]:
```suggestion
                bufferSize: BufferSize,
```

File: Jellyfin.Server/Infrastructure/SymlinkFollowingPhysicalFileResultExecutor.cs [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+                options: (AsyncFile.UseAsyncIO ? FileOptions.Asynchronous : FileOptions.None) | FileOptions.SequentialScan);
+
+            fileStream.Seek(offset, SeekOrigin.Begin);
+            await StreamCopyOperation
+                .CopyToAsync(fileStream, response.Body, count, bufferSize, CancellationToken.None)
```
Comment (cvium) [inline Jellyfin.Server/Infrastructure/SymlinkFollowingPhysicalFileResultExecutor.cs:139]:
```suggestion
                .CopyToAsync(fileStream, response.Body, count, BufferSize, CancellationToken.None)
```

**Review by crobibero — APPROVED** [review]
Confirmed the `Content-Length` header is the same for direct play (normal), and direct play (symlink)

**Comment by dekadius** [comment]
I'm wondering how this will behave with dangling symlinks. 

The same problem with streaming from symlinks existed in a 10.6 version, and a fix was implemented in [ManagedFileSystem.GetFileSystemMetaData](https://github.com/jellyfin/jellyfin/blob/master/Emby.Server.Implementations/IO/ManagedFileSystem.cs#L244). I imagine that fix is still needed for getting the correct file size to store in the db, but I see that newer versions of that fix handles dangling symlinks. I haven't been able to this PR yet, so this might be a false alarm, but catching the same exception and setting `Exists` to `false` in that case seems logical to me. Also it might be worth noting the difference in how the symlink file stream is opened.

Sorry if this already has been tested / considered.

**Comment by cvium** [comment]
I imagine it will fail, which is alright.

**Comment by random-human-being** [comment]
Can we expect a release with this fix before 10.8? A lot of content on my server is effectively inaccessible with the latest release.

**Comment by cvium** [comment]
Unfortunately not

**Pull request merged**